fe/cycles.py

```diff
@@ -0,0 +1,155 @@
+    -------
+    predict_fs : callable
+
+        predict_fs(simulated_rbfes, Optional[simulated_abfes]) -> fs
+
+        maps (an array containing n_rbfes *relative* calculations, and optionally
+              an array containing n_absolute *absolute* calculations)
+            -> an array containing n_nodes *absolute* estimates
+
+        this is a jax-transformable function, and in particular this means it can be used to define and differentiate
+        losses in terms of "cycle-closure-corrected" free energy estimates
+
+    Example
+    -------
+    >>> n_nodes = 3
+    >>> rbfe_inds = np.array([[0,1], [1,2], [2,0]])
+    >>> simulated_rbfes = np.array([-1.0, -1.0, +2.0])
+    >>> rbfe_sigmas = np.ones(len(rbfe_inds))
+    >>> predict_fs = construct_mle_layer(n_nodes, rbfe_inds, rbfe_sigmas)
+    >>> fs = predict_fs(simulated_rbfes, np.array([0.0]))
+    >>> reconstructed_diffs = onp.array([fs[j] - fs[i] for (i, j) in rbfe_inds])
+    >>> onp.isclose(reconstructed_diffs, simulated_rbfes).all()
+    True
+    >>> loss = lambda x : np.sum(predict_fs(x)**2) # random loss defined in terms of cycle-corrected estimates
+    >>> grad(loss)(simulated_rbfes) # can take gradients of this loss w.r.t. simulated_rbfes
+    DeviceArray([-2.66666667, -0.66666667,  3.33333333], dtype=float64)
+
+    Notes
+    -----
+    * Here we used an independent Gaussian noise model for the likelihood of simulated_rbfe(i, j) given
+        true underlying value of fs[j] - fs[i].
+        Other noise models could be plugged in here (e.g. ones that allow heavy-tailed noise or correlated errors),
+        as long as log_likelihood still permits a cvxpy-friendly expression.
+    * Here we used no prior information about plausible values of fs.
+        Rather than returning a maximum likelihood estimate
+            argmax_{trial_fs} log_likelihood(trial_fs),
+        we could just as well compute a maximum a posteriori estimate
+            argmax_{trial_fs} log_prior(trial_fs) + log_likelihood(trial_fs),
+        where log_prior(trial_fs) could be informed by a cheminformatics model or similar.
+
+    References
+    ----------
+    * DiffNet implementation
+        https://github.com/forcefield/DiffNet
+    * DiffNet paper
+        Huafeng Xu, Optimal measurement network of pairwise differences,
+        J. Chem. Inf. Model. 59, 4720-4728, 2019, https://doi.org/10.1021/acs.jcim.9b00528.
+    """
+
+    # get RBFE info
+    n_rbfes = len(rbfe_inds)
+    inds_l, inds_r = rbfe_inds.T
+    if (inds_l == inds_r).any():
+        raise AssertionError(f'invalid rbfe_inds -- includes self-comparisons: {rbfe_inds[(inds_l == inds_r)]}')
+
+    # check that the "map" is connected
+    if len(set(rbfe_inds.flatten())) != n_nodes:
+        present = set(rbfe_inds.flatten())
+        expected = set(np.arange(n_nodes))
+        missing_inds = list(expected.difference(present))
+        raise AssertionError(f'invalid rbfe_inds -- missing comparisons for nodes: {missing_inds}')
```

Comment:
Is this a hard requirement? I think for many datasets this won't be necessarily true.

Reply:
On second thought, I think we might need to meet an even more stringent condition than the one I actually checked here!

If we have `rbfe_inds = [[0,1], [2,3]]`, `simulated_rbfes = [1, 1]`, then any `fs = [0, 1, offset, offset + 1]` for arbitrary value of `offset` would equally well maximize the likelihood of `simulated_rbfes`. The information provided by `simulated_rbfes` provides no basis for comparing connected component `[0,1]` with connected component `[2,3]`, and I think we would not want to return an answer `fs` that could arbitrarily corrupt the between-component comparisons `fs[2] - fs[1]`, `fs[2] - fs[0]`, `fs[3] - fs[1]`, `fs[3] - fs[0]`.

To avoid this issue, I think it only makes sense to call this function on comparison networks containing a single connected component. In cases where the comparison network is disconnected, but we still wish to say something about comparisons _within_ each connected component, we should analyze each component independently, rather than altogether.

Reply:
Possibly addressed in https://github.com/proteneer/timemachine/pull/399/commits/7e4ca4890feb72638d814511e129e64ee282d322 ?

Reply:
Added a utility function to get connected components https://github.com/proteneer/timemachine/pull/399/commits/4d36492598730001dccc65e58a09f45910b7f4b9 .

If we have a bag of relative and absolute calculations that contains isolated "islands" (e.g. if we had two non-overlapping star maps, with no absolute calculations included), this function would tell us to analyze the two non-overlapping star maps independently. If we had at least one absolute calculation in each non-overlapping star map, then this function would tell us it's okay to analyze them jointly.